README.md
# surround_frequency_generator

A package to create surround sound effects

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.


CHANGELOG.md
## [1.0.0] - Oct 10, 2021

* Forking repository from surround_sound package.
* Updating flutter packages.
* Changing default waveform from SQUARE to SINUSOIDAL.
* Implementing functionality for iOS.

## [0.3.2] - Oct 10, 2021

* Update flutter packages

## [0.3.1] - April 28, 2021

* Expose the previous change in a separate function

## [0.3.0] - April 27, 2021

* Remove Tonejs dependencies from v0.2.x
* Remove ios bug

## [0.1.7] - March 12, 2021

* Upgrade flutter version
* Remove unwanted dependencies
* Upgrade packages

## [0.1.6] - November 27, 2020

* Webview version update

## [0.1.5] - September 02, 2020

* Minor bug fix in controller
* Add documentation to AudioParam

## [0.1.4] - August 31, 2020

* Sensitivity of x, y and z axis was increased

## [0.1.3] - July 31, 2020

* Bug fixes in html
* IOS bug fixes in Plist file (For Flutter WebView)

## [0.1.2] - July 17, 2020

* Now background colors can be set to the widget

## [0.1.1] - July 15, 2020

* Bug fixes and improvements

## [0.1.0] - July 13, 2020

* Minimum product that can be used successfully



example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:surround_frequency_generator/surround_frequency_generator.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Surround Sound Demo',
      home: HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  final _controller = SoundController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Surround Sound Example"),
      ),
      body: ListView(
        children: <Widget>[
          SoundWidget(
            soundController: _controller,
            backgroundColor: Colors.green,
          ),
          SizedBox(height: 32),
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              MaterialButton(
                child: Text("Play"),
                onPressed: () async {
                  await _controller.play();
                  final val = await _controller.isPlaying();
                  print('isPlaying: $val');
                },
              ),
              SizedBox(width: 24),
              MaterialButton(
                child: Text("Stop"),
                onPressed: () async {
                  await _controller.stop();
                  final val = await _controller.isPlaying();
                  print('isPlaying: $val');
                },
              ),
            ],
          ),
          SizedBox(height: 32),
          ValueListenableBuilder<AudioParam>(
            valueListenable: _controller,
            builder: (context, value, _) {
              return Column(
                children: <Widget>[
                  Text("Volume"),
                  Slider(
                    value: value.volume,
                    min: 0,
                    max: 1,
                    onChanged: (val) {
                      _controller.setVolume(val);
                    },
                  ),
                  Text("Frequency"),
                  Slider(
                    value: value.freq,
                    min: 128,
                    max: 1500,
                    onChanged: (val) {
                      _controller.setFrequency(val);
                    },
                  ),
                  SizedBox(height: 32),
                  Text(
                    "Position",
                    style: TextStyle(
                      fontWeight: FontWeight.w600,
                      fontSize: 18,
                    ),
                  ),
                  SizedBox(height: 12),
                  Text("x-axis"),
                  Slider(
                    value: value.x,
                    min: -0.2,
                    max: 0.2,
                    onChanged: (val) {
                      _controller.setPosition(val, value.y, value.z);
                    },
                  ),
                  Text("y-axis"),
                  Slider(
                    value: value.y,
                    min: -0.2,
                    max: 0.2,
                    onChanged: (val) {
                      _controller.setPosition(value.x, val, value.z);
                    },
                  ),
                  Text("z-axis"),
                  Slider(
                    value: value.z,
                    min: -0.2,
                    max: 0.2,
                    onChanged: (val) {
                      _controller.setPosition(value.x, value.y, val);
                    },
                  ),
                ],
              );
            },
          ),
        ],
      ),
    );
  }
}


lib/src/web_html.dart
String html(String background) {
  return '''
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Audio API</title>
</head>
<body style="background-color:$background;">
<script>
    console.clear();
    let audioCtx;
    let panner;
    let listener;
    let oscillator;
    let volumeController;

    let started = false;
    let xPos = 30;
    let yPos = 30;
    let zPos = 300;
    let freq = 512;
    let vol = 0.5;

    function _setPanner() {
        if (panner.positionX) {
            panner.positionX.value = xPos;
            panner.positionY.value = yPos;
            panner.positionZ.value = zPos;
        } else {
            panner.setPosition(xPos, yPos, zPos);
        }
    }

    async function _things() {
        let AudioContext = window.AudioContext || window.webkitAudioContext;
        audioCtx = new AudioContext();

        panner = audioCtx.createPanner();
        panner.panningModel = 'HRTF';
        panner.distanceModel = 'inverse';
        panner.refDistance = 1;
        panner.maxDistance = 10000;
        panner.rolloffFactor = 1;
        panner.coneInnerAngle = 360;
        panner.coneOuterAngle = 0;
        panner.coneOuterGain = 0;

        if (panner.orientationX) {
            panner.orientationX.value = 0;
            panner.orientationY.value = 0;
            panner.orientationZ.value = -1;
        } else {
            panner.setOrientation(0, 0, -1);
        }

        listener = audioCtx.listener;

        if (listener.forwardX) {
            listener.forwardX.value = 0;
            listener.forwardY.value = 0;
            listener.forwardZ.value = 1;
            listener.upX.value = 0;
            listener.upY.value = 1;
            listener.upZ.value = 0;
        } else {
            listener.setOrientation(0, 0, 1, 0, 1, 0);
        }

        if (listener.positionX) {
            listener.positionX.value = xPos;
            listener.positionY.value = yPos;
            listener.positionZ.value = zPos;
        } else {
            listener.setPosition(xPos, yPos, zPos);
        }

        _setPanner();

        oscillator = audioCtx.createOscillator();
        oscillator.type = 'sinusoidal';
        // oscillator.type = 'square';
        oscillator.frequency.value = freq;
        oscillator.start();

        volumeController = audioCtx.createGain();
        volumeController.gain.value = vol;

        // Connecting
        oscillator.connect(panner);
        panner.connect(volumeController)
    }

    async function init_sound() {
        if (!audioCtx) {
            await _things();
        }

        // check if context is in suspended state (autoplay policy)
        if (audioCtx.state === 'suspended') {
            audioCtx.resume();
        }
    }

    function play() {
        if (!started) {
            volumeController.connect(audioCtx.destination);
            started = true;
        }
    }

    function stop() {
        if (started) {
            volumeController.disconnect(audioCtx.destination);
            started = false;
        }
    }

    function set_panner(x, y, z) {
        xPos = x;
        yPos = y;
        zPos = z;
        _setPanner();
    }

    function set_freq(f) {
        freq = f;
        oscillator.frequency.value = freq;
    }
    
    function force_set_freq(f) {
        freq = f;
        
        volumeController.disconnect(audioCtx.destination);
        started = false;
        oscillator.stop();
        
        oscillator = audioCtx.createOscillator();
        oscillator.type = 'square';
        oscillator.frequency.value = freq;
        oscillator.start();
        
        // Connecting
        oscillator.connect(panner);
        panner.connect(volumeController)
        volumeController.connect(audioCtx.destination);
        started = true;
    }

    function set_volume(v) {
        vol = v;
        volumeController.gain.value = vol;
    }

</script>
</body>
</html>
''';
}


lib/src/audio_param.dart
import 'package:flutter/widgets.dart';

/// {@template audio_param}
/// Data model that can be used to transfer data between the app UI
/// and the sound source. Use with the [SoundController].
/// {@endtemplate}
@immutable
class AudioParam {
  /// Value from -1 to 1 where -1 being right side and 1 being left side
  final double x;

  /// Value from -1 to 1 where -1 being bottom and 1 being top
  final double y;

  /// Value from -1 to 1 where -1 being back and 1 being front
  final double z;

  /// Value from 20 to 20000 which controls the frequency of the
  /// sound source
  final double freq;

  /// Value from 0 to 1 which controls the volume of the sound
  /// source where 0 being mute and 1 being highest volume.
  final double volume;

  /// {@macro audio_param}
  AudioParam({
    this.x = 0,
    this.y = 0,
    this.z = 0,
    this.freq = 512,
    this.volume = 0.5,
  });

  /// Create a copy of the [AudioParam] with the given non-null values.
  AudioParam copyWith({
    double? x,
    double? y,
    double? z,
    double? freq,
    double? volume,
  }) {
    return AudioParam(
      x: x ?? this.x,
      y: y ?? this.y,
      z: z ?? this.z,
      freq: freq ?? this.freq,
      volume: volume ?? this.volume,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is AudioParam &&
          runtimeType == other.runtimeType &&
          x == other.x &&
          y == other.y &&
          z == other.z &&
          freq == other.freq &&
          volume == other.volume;

  @override
  int get hashCode =>
      x.hashCode ^ y.hashCode ^ z.hashCode ^ freq.hashCode ^ volume.hashCode;
}


lib/src/sound_widget.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:surround_frequency_generator/src/sound_controller.dart';
import 'package:surround_frequency_generator/src/web_html.dart';
import 'package:webview_flutter/webview_flutter.dart';
import 'package:webview_flutter_wkwebview/webview_flutter_wkwebview.dart';
import 'package:webview_flutter_platform_interface/webview_flutter_platform_interface.dart';

class SoundWidget extends StatefulWidget {

  final SoundController soundController;
  final Color backgroundColor;
  final WebViewController? webViewAndroidController;
  final PlatformWebViewController? webViewIosController;

  SoundWidget({
    Key? key,
    required this.soundController,
    this.backgroundColor = Colors.white,
    this.webViewAndroidController,
    this.webViewIosController,
  }) : super(key: key);

  @override
  _SoundWidgetState createState() => _SoundWidgetState();
}

class _SoundWidgetState extends State<SoundWidget> {
  late String htmlText;
  late WebViewController webViewAndroidController;

  PlatformWebViewCookieManager? cookieManager;
  late PlatformWebViewController webViewIosController;

  @override
  void initState() {
    super.initState();
    if(widget.webViewAndroidController != null) {
      webViewAndroidController = widget.webViewAndroidController!;
    } else if(widget.webViewIosController != null) {
      webViewIosController = widget.webViewIosController!;
    }

    _init();
  }

  @override
  void dispose() {
    if(Platform.isAndroid) {
      webViewAndroidController.clearCache(); // Clear the WebView cache (optional)
      webViewAndroidController.goBack();    // Dispose of the WebView
    } else {
      webViewIosController.clearCache();
      webViewIosController.goBack();
    }

    super.dispose();
  }

  void _init() {

    final color = widget.backgroundColor.value.toRadixString(16).substring(2);
    final String htmlBase64 = base64Encode(
      const Utf8Encoder().convert(html('#$color')),
    );

    if(Platform.isAndroid) {
      setWebViewForAndroid(Uri.parse('data:text/html;base64,$htmlBase64'));
    } else {
      setWebViewForIOS(Uri.parse('data:text/html;base64,$htmlBase64'));
    }

  }

  @override
  void didUpdateWidget(SoundWidget oldWidget) {
    super.didUpdateWidget(oldWidget);
    if (widget.backgroundColor != oldWidget.backgroundColor) {
      _init();
    }
  }

  @override
  Widget build(BuildContext context) {
    return SizedBox(
      height: 1,
      width: 1,
      child: Platform.isAndroid ? WebViewWidget(
        controller: webViewAndroidController,
      ) : PlatformWebViewWidget(
        PlatformWebViewWidgetCreationParams(controller: webViewIosController),
      ).build(context),
    );
  }

  void setWebViewForAndroid(Uri uri) {

    // #docregion platform_features
    late final PlatformWebViewControllerCreationParams params;
    if (WebViewPlatform.instance is WebKitWebViewPlatform) {
      params = WebKitWebViewControllerCreationParams(
        allowsInlineMediaPlayback: true,
        mediaTypesRequiringUserAction: const <PlaybackMediaTypes>{},
      );
    } else {
      params = const PlatformWebViewControllerCreationParams();
    }

    webViewAndroidController = WebViewController.fromPlatformCreationParams(params);
    webViewAndroidController.setJavaScriptMode(JavaScriptMode.unrestricted);
    // #enddocregion platform_features
    webViewAndroidController.loadRequest(uri);
    webViewAndroidController.setNavigationDelegate(
        NavigationDelegate(
          onProgress: (int progress) {
            print('WebView is loading (progress : $progress%)');
          },
          onPageStarted: (String url) {
            widget.soundController.complete(androidController: webViewAndroidController);
            print('Page started loading: $url');
          },
          onPageFinished: (String url) {
            webViewAndroidController.runJavaScript('init_sound();');
            print('Page finished loading: $url');
          },
        )
    );
  }

  void setWebViewForIOS(Uri uri) {

    webViewIosController= PlatformWebViewController(
      WebKitWebViewControllerCreationParams(allowsInlineMediaPlayback: true),
    )
      ..setJavaScriptMode(JavaScriptMode.unrestricted)
      ..setPlatformNavigationDelegate(
        PlatformNavigationDelegate(
          const PlatformNavigationDelegateCreationParams(),
        )
          ..setOnProgress((int progress) {
            print('WebView is loading (progress : $progress%)');
          })
          ..setOnPageStarted((String url) {
            widget.soundController.complete(iOSController: webViewIosController);
            print('Page started loading: $url');
          })
          ..setOnPageFinished((String url) {
            webViewIosController.runJavaScript('init_sound();');
            print('Page finished loading: $url');
          })
          ..setOnWebResourceError((WebResourceError error) {
            print('''
              Page resource error:
              code: ${error.errorCode}
              description: ${error.description}
              errorType: ${error.errorType}
              isForMainFrame: ${error.isForMainFrame}
              url: ${error.url}
            ''');
          })
      )
      ..addJavaScriptChannel(JavaScriptChannelParams(
        name: 'Toaster',
        onMessageReceived: (JavaScriptMessage message) {
          ScaffoldMessenger.of(context).showSnackBar(
            SnackBar(content: Text(message.message)),
          );
        },
      ))
      ..setOnPlatformPermissionRequest(
            (PlatformWebViewPermissionRequest request) {
          debugPrint(
            'requesting permissions for ${request.types.map((WebViewPermissionResourceType type) => type.name)}',
          );
          request.grant();
        },
      )
      ..loadRequest(LoadRequestParams(uri: uri),);
  }

}


lib/src/sound_controller.dart
import 'dart:async';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:webview_flutter/webview_flutter.dart';
import 'package:webview_flutter_platform_interface/webview_flutter_platform_interface.dart';

import 'audio_param.dart';

class SoundController extends ValueNotifier<AudioParam> {

  WebViewController? _webViewAndroidController;
  Completer<WebViewController> _webAndroidController = Completer<WebViewController>();
  PlatformWebViewController? _webViewIOSController;
  Completer<PlatformWebViewController> _webIOSController = Completer<PlatformWebViewController>();

  SoundController() : super(AudioParam());

  set value(param) {
    setPosition(param.x, param.y, param.z);
    setFrequency(param.freq);
    setVolume(param.volume);
    super.value = param;
  }

  void complete({WebViewController? androidController, PlatformWebViewController? iOSController}) {
    if(Platform.isAndroid && androidController != null) {
      if (_webAndroidController.isCompleted) {
        _webAndroidController = Completer<WebViewController>();
        _webViewAndroidController = null;
      }
      _webAndroidController.complete(androidController);
    } else if(iOSController != null) {
      if (_webIOSController.isCompleted) {
        _webIOSController = Completer<PlatformWebViewController>();
        _webViewIOSController = null;
      }
      _webIOSController.complete(iOSController);
    }
  }

  Future init() async {

    try {
      if(Platform.isAndroid) {
        _webViewAndroidController = await _webAndroidController.future;
        await _webViewAndroidController!.runJavaScript('init_sound();');
      } else {
        _webViewIOSController = await _webIOSController.future;
        await _webViewIOSController!.runJavaScript('init_sound();');
      }
    } catch (e) {
      print(e.toString());
    }
  }

  Future _check() async {
    if(Platform.isAndroid) {
      if (_webViewAndroidController == null) {
        await init();
      }
    } else {
      if (_webViewIOSController == null) {
        await init();
      }
    }

  }

  Future play() async {
    await _check();
    if(Platform.isAndroid) {
      await _webViewAndroidController?.runJavaScript('play();');
    } else {
      await _webViewIOSController?.runJavaScript('play();');
    }
  }

  Future stop() async {
    await _check();
    if(Platform.isAndroid) {
      await _webViewAndroidController?.runJavaScript('stop();');
    } else {
      await _webViewIOSController?.runJavaScript('stop();');
    }
  }

  Future<bool> isPlaying() async {
    try {
      await _check();
      if(Platform.isAndroid) {
        final started = await _webViewAndroidController?.runJavaScriptReturningResult('started');
        if(started is bool) {return started;}
      } else {
        final started = await _webViewIOSController?.runJavaScriptReturningResult('started');
        if(started is bool) {return started;}
      }
    } catch (e) {
      print(e.toString());
    }

    return false;
  }

  Future setPosition(double x, double y, double z) async {
    await _check();
    x = x.clamp(-1.0, 1.0);
    y = y.clamp(-1.0, 1.0);
    z = z.clamp(-1.0, 1.0);
    super.value = super.value.copyWith(x: x, y: y, z: z);

    if(Platform.isAndroid) {
      await _webViewAndroidController?.runJavaScript('set_panner('
          '${x * 5.5 + 30}, '
          '${y * 5.5 + 30}, '
          '${z * 5.5 + 300}'
          ');');
    } else {
      await _webViewIOSController?.runJavaScript('set_panner('
          '${x * 5.5 + 30}, '
          '${y * 5.5 + 30}, '
          '${z * 5.5 + 300}'
          ');');
    }
  }

  Future forceSetFrequency(double freq) async {
    await _check();
    freq = freq.clamp(20.0, 20000.0);
    super.value = super.value.copyWith(freq: freq);

    if(Platform.isAndroid) {
      await _webViewAndroidController?.runJavaScript('force_set_freq($freq);');
    } else {
      await _webViewIOSController?.runJavaScript('force_set_freq($freq);');
    }

  }

  Future setFrequency(double freq) async {
    await _check();
    freq = freq.clamp(20.0, 20000.0);
    super.value = super.value.copyWith(freq: freq);
    if(Platform.isAndroid) {
      await _webViewAndroidController?.runJavaScript('set_freq($freq);');
    } else {
      await _webViewIOSController?.runJavaScript('set_freq($freq);');
    }
  }

  Future setVolume(double vol) async {
    await _check();
    vol = vol.clamp(0.0, 1.0);
    super.value = super.value.copyWith(volume: vol);
    if(Platform.isAndroid) {
      await _webViewAndroidController?.runJavaScript('set_volume($vol);');
    } else {
      await _webViewIOSController?.runJavaScript('set_volume($vol);');
    }
  }

}
